pkgs/development/compilers/flutter/wrapper.nix

```diff
@@ -32,30 +44,75 @@
 , clang
+, lndir
+, symlinkJoin
 }:
 
 let
-  flutterWithArtifacts = flutter.override {
-    includedEngineArtifacts = if includedEngineArtifacts != null then includedEngineArtifacts else {
-      common = [
-        "flutter_patched_sdk"
-        "flutter_patched_sdk_product"
-      ];
-      platform = {
-        android = lib.optionalAttrs supportsAndroid
-          ((lib.genAttrs [ "arm" "arm64" "x64" ] (architecture: [ "profile" "release" ])) // { x86 = [ "jit-release" ]; });
-        linux = lib.optionalAttrs supportsLinuxDesktop
-          (lib.genAttrs ((lib.optional stdenv.hostPlatform.isx86_64 "x64") ++ (lib.optional stdenv.hostPlatform.isAarch64 "arm64"))
-            (architecture: [ "debug" "profile" "release" ]));
-      };
-    };
+  engineArtifactDirectory =
+    let
+      engineArtifacts = callPackage ./engine-artifacts { inherit (flutter) engineVersion; };
+    in
+    runCommandLocal "flutter-engine-artifacts-${flutter.version}" { }
+      (
+        let
+          mkCommonArtifactLinkCommand = { artifact }:
+            ''
+              mkdir -p $out/artifacts/engine/common
+              ${lndir}/bin/lndir -silent ${artifact} $out/artifacts/engine/common
+            '';
+          mkPlatformArtifactLinkCommand = { artifact, os, architecture, variant ? null }:
+            let
+              artifactDirectory = "${os}-${architecture}${lib.optionalString (variant != null) "-${variant}"}";
+            in
+            ''
+              mkdir -p $out/artifacts/engine/${artifactDirectory}
+              ${lndir}/bin/lndir -silent ${artifact} $out/artifacts/engine/${artifactDirectory}
+            '';
+        in
+        builtins.concatStringsSep "\n"
+          ((map
+            (name: mkCommonArtifactLinkCommand {
+              artifact = engineArtifacts.common.${name};
+            })
+            (includedEngineArtifacts.common or [ ])) ++
+          (builtins.foldl'
+            (commands: os: commands ++
+              (builtins.foldl'
+                (commands: architecture: commands ++
+                  (builtins.foldl'
+                    (commands: variant: commands ++
+                      (map
+                        (artifact: mkPlatformArtifactLinkCommand {
+                          inherit artifact os architecture variant;
+                        })
+                        engineArtifacts.platform.${os}.${architecture}.variants.${variant}))
+                    (map
+                      (artifact: mkPlatformArtifactLinkCommand {
+                        inherit artifact os architecture;
+                      })
+                      engineArtifacts.platform.${os}.${architecture}.base)
+                    includedEngineArtifacts.platform.${os}.${architecture}))
+                [ ]
+                (builtins.attrNames includedEngineArtifacts.platform.${os})))
+            [ ]
+            (builtins.attrNames (includedEngineArtifacts.platform or { }))))
+      );
+
+  cacheDir = symlinkJoin {
```

Comment:
We know that engine artifacts cannot be symlinked. Are we sure that other files in `bin/cache` can?

Reply:
I don't really understand, could you clarify what you mean? 

The issue is that flutter always looks into it's bin/cache where the `bin` directory is where the flutter executable itself is located. So without copying the executable and everything around it, it cannot be changed.

I think symlinking the artifacts and other cache files together shouldn't cause any issues as long as we make flutter consider the derivation its cache.

Reply:
Ah, my bad. I got confused.